pubsub/mqtt/mqtt.go

```diff
@@ -324,20 +313,61 @@ func (m *mqttPubSub) handlerForTopic(topic string) pubsub.Handler {
 	return nil
 }
 
-func (m *mqttPubSub) connect(ctx context.Context, clientID string) (mqtt.Client, error) {
+func (m *mqttPubSub) connect(ctx context.Context, clientID string, isConsumer bool, readyCh chan<- struct{}) (mqtt.Client, error) {
 	uri, err := url.Parse(m.metadata.url)
 	if err != nil {
 		return nil, err
 	}
 	opts := m.createClientOptions(uri, clientID)
-	// Turn off auto-ack
-	opts.SetAutoAckDisabled(true)
-	client := mqtt.NewClient(opts)
+	var client mqtt.Client
+
+	// For consumers, add the topics we're subscribing to in the OnConnectHandler
+	if isConsumer {
+		sendNotification := func() {
+			if readyCh != nil {
+				close(readyCh)
+				readyCh = nil
+			}
+		}
+		fn := m.onMessage(ctx)
+		opts.SetOnConnectHandler(func(client mqtt.Client) {
+			if len(m.topics) == 0 {
+				sendNotification()
+				return
+			}
+
+			// Subscribe to topics upon connecting
+			subscribeTopics := make(map[string]byte, len(m.topics))
+			for k := range m.topics {
+				subscribeTopics[k] = m.metadata.qos
+			}
 
-	// Add all routes before we connect to catch messages that may be delivered before client.Subscribe is invoked
-	// The routes will be overwritten later
-	for topic := range m.topics {
-		client.AddRoute(topic, m.onMessage(ctx))
+			token := client.SubscribeMultiple(subscribeTopics, fn)
+			ok := token.WaitTimeout(defaultWait)
+			tokenErr := token.Error()
+			// In case of timeout, ok will be false and there won't be an error in the token
+			if tokenErr == nil && !ok {
+				tokenErr = context.DeadlineExceeded
+			}
+			if tokenErr != nil {
+				// Cause a disconnection and wait for reconnection
+				m.logger.Errorf("MQTT error while trying to subscribe to topics; will reconnect. Error: %v", tokenErr)
+				client.Disconnect(5)
+				m.connect(ctx, clientID, isConsumer, readyCh)
```

Comment:
This would essentially lead to recursive calls to connect, and indefinite retries in case of subscribe failures. Is this expected behavior?

Reply:
I would say so, until it recovers. There’s a timeout in starting subscriptions anyways.

Reply:
Should we have some control in terms of number of retries? Let's say, if it doesn't timeout but fails due to some non-retriable reason, it will just keep on re-trying. If we can't differentiate between retriable and non-retriable errors here, may be having limited number of attempts may help.

Reply:
Let me look at how it can be improved

Reply:
@ItalyPaleAle anything remaining here?